src/Microsoft.Restier.Core/ApiConfiguration.cs

```diff
@@ -65,5 +68,46 @@ public IServiceProvider ServiceProvider
         {
             return this.serviceProvider.GetService<T>();
         }
+
+        internal TaskCompletionSource<IEdmModel> CompeteModelGeneration(out Task<IEdmModel> running)
+        {
+            var source = new TaskCompletionSource<IEdmModel>(TaskCreationOptions.AttachedToParent);
+            var exist = Interlocked.CompareExchange(ref modelTask, source.Task, null);
+            if (exist != null)
+            {
+                running = exist;
+                source.SetCanceled();
+                return null;
+            }
+
+            source.Task.ContinueWith(
+                task => FinishModelGeneration(task),
+                TaskContinuationOptions.ExecuteSynchronously);
+            running = null;
+            return source;
+        }
+
+        private void FinishModelGeneration(Task<IEdmModel> task)
+        {
+            if (task.Status == TaskStatus.RanToCompletion)
+            {
+                ReplaceModelTask(Task.FromResult(task.Result), task);
```

Comment:
Partially, the main purpose is to confirm that this is truly the original model generation task. But actually it must be. In a previous version FinishModelGeneration is called from GetModelAsync, so I added this validation.
And I didn't look into Task code but intuitively I think a task built from direct result would consume less resources than a completed task from a completion source.